Desktop,Mobile,Cli: Fixes #11017: Delete revisions on the sync target when deleted locally

personalizedrefrigerator opened the pull request:
# Summary

This pull request fixes #11017 &mdash; previously, old revisions would remain on the sync target, even after being deleted locally. It seems that these deleted revisions [could also be downloaded on sync](https://github.com/laurent22/joplin/issues/11017#issuecomment-2347154020) soon after being deleted.

# Testing plan

**Manual testing**: I've done only limited manual regression testing:
- (Desktop/Fedora 40): Verified that synchronization starts completes successfully with a local Joplin Server/Cloud instance.

**Automated testing**: This pull request adds new automated tests:
- `'should delete old revisions remotely when deleted locally'`: Verifies that revision deletions sync between clients and that new clients don't download deleted revisions.
- `'should not error on revisions for missing (not downloaded yet/permanently deleted) notes'`: Verifies that old revisions are correctly deleted even if the corresponding item no longer exists.


<!--

Please prefix the title with the platform you are targetting:

Here are some examples of good titles:

- Desktop: Resolves #123: Added new setting to change font
- Mobile, Desktop: Fixes #456: Fixed config screen error
- All: Resolves #777: Made synchronisation faster

And here's an explanation of the title format:

- "Desktop" for the Windows/macOS/Linux app (Electron app)
- "Mobile" for the mobile app (or "Android" / "iOS" if the pull request only applies to one of the mobile platforms)
- "CLI" for the CLI app

If it's two platforms, separate them with commas - "Desktop, Mobile" or if it's for all platforms, prefix with "All".

If it's not related to any platform (such as a translation, change to the documentation, etc.), simply don't add a platform.

Then please append the issue that you've addressed or fixed. Use "Resolves #123" for new features or improvements and "Fixes #123" for bug fixes.

AND PLEASE READ THE GUIDE: https://github.com/laurent22/joplin/blob/dev/readme/dev/index.md

-->

personalizedrefrigerator commented on packages/lib/models/Revision.ts:
```diff
@@ -305,43 +308,82 @@ export default class Revision extends BaseItem {
 
-			const keptRev = await this.modelSelectOne('SELECT * FROM revisions WHERE item_updated_time >= ? AND item_type = ? AND item_id = ? ORDER BY item_updated_time ASC LIMIT 1', [cutOffDate, rev.item_type, rev.item_id]);
+			if (readOnly) {
+				this.logger().debug('Not deleting revisions for read-only shared item. ID:', itemId);
```
I wonder if this should be a `logger.info`? A `.info` could potentially lead to a very large number of messages on each `deleteOldRevisions` call.

personalizedrefrigerator commented on packages/lib/models/Revision.ts:
```diff
@@ -305,43 +308,82 @@ export default class Revision extends BaseItem {
-			const doneKey = `${rev.item_type}_${rev.item_id}`;
-			if (doneItems[doneKey]) continue;
+		const deleteOldRevisionsForItem = async (itemType: ModelType, itemId: string, oldRevisions: RevisionEntity[]) => {
+			const item = await BaseItem.loadItemById(itemId);
```
- [x] Verify that this doesn't fail if:
      1. A note is created.
      2. The note is permanently deleted.
      3. The note's revisions are processed.

laurent22 commented:
The diff is a bit hard to follow. Do you know what was special about revisions which means that they wouldn't be deleted after sync?

mrjo118 commented:
> The diff is a bit hard to follow. Do you know what was special about revisions which means that they wouldn't be deleted after sync?

From what I understand, there were no deleted_items table entries created when deleting the revisions. But the new call to this.batchDelete will create some on all sync sources. I don't follow what all the other changes are for though

personalizedrefrigerator commented:
> From what I understand, there were no deleted_items table entries created when deleting the revisions. But the new call to this.batchDelete will create some on all sync sources.

That's correct! The other changes include:
- Refactoring to simplify calling `batchDelete`.
- Removed the SQL query that deleted revisions previously — it has been replaced by `batchDelete`.
- Added more tests.
- Added logic to prevent revisions for items in a read-only share from being deleted.
   - As commented above, this change needs more testing: It might break if revisions still exist after the original item has been permanently deleted.

personalizedrefrigerator commented:
> Added logic to prevent revisions for items in a read-only share from being deleted. 

Currently, this is done by loading the original item (with `BaseItem.loadItemById`) and checking whether it's part of a read only share. 

This could be problematic: If the original item doesn't exist (e.g. not downloaded yet or permanently deleted), then the revisions can be deleted anyway. I'm not sure how to prevent this &mdash; would checking if the parent folder for the revisions is part of a read-only share (with `RevisionEntity.parent_id`) be sufficient? Would it be possible/make sense to add a `share_id` property to revisions?

**Update**: Answered [here](https://joplinapp.org/help/dev/spec/read_only/#at-the-sync-level).

personalizedrefrigerator commented on packages/lib/models/Revision.ts:
```diff
@@ -305,43 +308,92 @@ export default class Revision extends BaseItem {
+		}
 
-			const keptRev = await this.modelSelectOne('SELECT * FROM revisions WHERE item_updated_time >= ? AND item_type = ? AND item_id = ? ORDER BY item_updated_time ASC LIMIT 1', [cutOffDate, rev.item_type, rev.item_id]);
+		const canDeleteRevisionsForItem = async (itemType: ModelType, itemId: string) => {
```
Note from meeting: Revisions aren't shared, so this isn't necessary.
